### ISSUE TITLE
Notifications not clearing after sync

### ISSUE THREAD

**Issue by benparsons** [vector-im/element-android#4862]
### Steps to reproduce

1. I get messages
2. Android shows me a notification
3. I read the messages on element-web
4. I get more messages
5. Android updates the notification to show me, **but those messages I already read are still shown in the notification**

#### What did you expect?

Read messages are cleared from notifications

#### What happened instead?

They remain.

### Your phone model

Samsung S20FE

### Operating system version

Android 11

### Application version and app store

1.3.12

### Will you send logs?

No

**Comment by ouchadam** [comment]
does this happen for all rooms/chats?

I've seen this issue happen in the past when the _message read_ push events are not sent due to push rules, I was able to get them working again by setting the room notifications to none and back to all

**Comment by benparsons** [comment]
I don't think this is uncommon, multiple customers and IRL friends report it, and I see it as basically my normal experience with the android app.

You could be right, that this is caused by push rules not being correctly set, next time I see this on a specific room I will toggle the notification settings for that room and see if it helps.

**Comment by ouchadam** [comment]
> I don't think this is uncommon, multiple customers and IRL friends report it, and I see it as basically my normal experience with the android app.

ah wasn't aware of more reports, happy to bump it up

**Comment by benparsons** [comment]
Naturally, now I've just tested this to see if I could try the push-rules question, I couldn't repo at all..! So, I will try to get a better repro and come back with more info.

**Comment by ouchadam** [comment]
if possible that would be great! the push flows are quite tricky as there's so many moving parts (firebase cloud messaging tokens, google play services availability, OEM battery optimisations, cross client push rules config, sygnal and the /sync)

**Comment by abmantis** [comment]
I've been also having this issue frequently now. Notifications arrive on time, but then most of the times they stay on the phone until I open Element. Once in a while. it works as expected and the notification is cleared automatically, but most of the times it is not.

**Comment by hho** [comment]
This has been an issue for me recently as well - I think it started in January this year (it used to work perfectly before).

- Pixel 6
- Element Android Beta
- hosting my own Synapse; never more than a week behind release of Debian packages

**Comment by BloodyIron** [comment]
This happens literally 100% of the time for me across all android devices I own. So I would not say "frequent" I would say "always" for frequency.

**Comment by f35f0ef9d0e827dae86552d3899f78fc** [comment]
I would like to add that this is not isolated to Android; the same behavior folks have been describing here has been happening on iOS for quite a while (a few months?) as well.

**Comment by SpiritCroc** [comment]
For those issues with notifications not dismissing when read on another client *until* opening Element Android, this should be the fix on synapse: https://github.com/matrix-org/synapse/pull/12721
This does not address the original issue however, in particular, if at "4. I get more messages" notifications aren't dismissed, there's something else broken as far as I see.

**Comment by BloodyIron** [comment]
ETA before that hits production public matrix servers?

**Comment by hho** [comment]
@SpiritCroc Thank you so much!! I monkey-patched my Synapse installation with the line from your PR and it works beautifully again!

**Comment by SpiritCroc** [comment]
> @SpiritCroc Thank you so much!! I monkey-patched my Synapse installation with the line from your PR and it works beautifully again!

Great, thanks for testing and reporting!

**Comment by wodny** [comment]
On Android (Element) notifications from the matrix.org instance disappear as excepted (again) after a message is read in web Element app. Thanks.

**Comment by CyberShadow** [comment]
I'm seeing notifications from keywords not clearing with Element-Android 1.4.27 and Synapse v1.60.0.

Notifications for things like private messages are being cleared.

The notification stays even in I open Element-Android, and only goes away when I open the room that the notification was generated in.

Is it possible that there's a similar bug with sending pushes for clearing keyword notifications?

**Comment by BloodyIron** [comment]
So this "fix" has hit production, and initially it improved things, but now Element Android does not seem to be lazy-loading when a device is idle/low power, and again notifications are not being cleared until a device starts getting used again... It seems one solution leads to another problem :O

**Comment by CyberShadow** [comment]
Still seeing this in Element-Android 1.6.12.

**Comment by CyberShadow** [comment]
That's surprising, as SchildiChat is an Element fork, so I'd expect it to inherit Element's bugs.

https://github.com/SchildiChat/SchildiChat-android/blob/sc/FEATURES.md lists two small notification-related features, but nothing about reliability.

Are you perhaps using SchildiChat Next (which is very different)?

**Comment by abmantis** [comment]
> That's surprising, as SchildiChat is an Element fork, so I'd expect it to inherit Element's bugs.
> 
> https://github.com/SchildiChat/SchildiChat-android/blob/sc/FEATURES.md lists two small notification-related features, but nothing about reliability.
> 
> Are you perhaps using SchildiChat Next (which is very different)?

I also have the same experience as @BloodyIron . On multiple occasions I've switched to SchildiChat (not Next) because of notification issues.

**Comment by SpiritCroc** [comment]
As far as I remember I have upstreamed all notification reliability fixes I did for SchildiChat, so don't know why it would differ from Element right now

**Comment by BloodyIron** [comment]
> That's surprising, as SchildiChat is an Element fork, so I'd expect it to inherit Element's bugs.
> 
> https://github.com/SchildiChat/SchildiChat-android/blob/sc/FEATURES.md lists two small notification-related features, but nothing about reliability.
> 
> Are you perhaps using SchildiChat Next (which is very different)?

Well proof's in the pudding, it's been head and shoulders more reliable for me and multiple other people around me than the Element Android app. Software can regress by the way, which is the case with Element frankly. Like there have been sooooooo many notification regressions, bugs, and problems over the years with Element alone. I previously chalked it up to "cutting their teeth" growing pain stuff, but at this point, it's just straight up unacceptably unreliable.

From a motivations perspective it's extremely surprising that notification reliability is not one of the top priorities a _chat_ app cares about.

**Comment by BloodyIron** [comment]
> As far as I remember I have upstreamed all notification reliability fixes I did for SchildiChat, so don't know why it would differ from Element right now

Your efforts are appreciated, thanks!

**Comment by CyberShadow** [comment]
> As I'm testing both at the same time, schildichat so far is worlds more reliable for receiving, sending, clearing, and other notification stuff that I can't remember this moment.

I can confirm that SchildiChat does not have the stuck notifications problem!

After I clear the notifications on my PC, they disappear from SchildiChat after a minute or so.

> As far as I remember I have upstreamed all notification reliability fixes I did for SchildiChat, so don't know why it would differ from Element right now

Interesting - I'm wondering if this could somehow be related to a difference in the push mechanism. Could it be something like Element and Element X using GCM (and Synapse having bugs with it), but SchildiChat using polling (which is less efficient but also not having push bugs)?

**Comment by hho** [comment]
@Atemu This issue (and project) is not about Element X at all.  
Please see https://github.com/element-hq/element-x-android/issues instead.

Element (no X) for Android is unsupported besides security fixes, so essentially there's no point contributing anything to this project anymore.

**Comment by Atemu** [comment]
OH. I did not realise I was in the element repo.

Sorry for the noise then; will report in the Element X repo instead.
